src/main/java/com/google/javascript/clutz/DeclarationGenerator.java

```diff
@@ -256,19 +266,33 @@ public String produceDts(Compiler compiler, Depgraph depgraph) {
       // parent symbol.
       if (symbol.getName().contains(".prototype")) continue;
 
-      String namespace = getNamespace(symbol.getName());
+      String parentPath = getNamespace(symbol.getName());
       boolean isDefault = isDefaultExport(symbol);
-      boolean isTypedef = symbol.getType() != null && isTypedef(symbol.getType());
-      // Typedef symbols do not enumerate their namespace properties.
-      if (!isDefault && !isTypedef) visitedNamespaces.add(symbol.getName());
-      if (isDefault && visitedNamespaces.contains(namespace)) continue;
-      declareNamespace(isDefault ? namespace : symbol.getName(), symbol, isDefault, compiler, transitiveProvides, true);
+
+      if (visitedNamespaces.contains(parentPath)) {
+        // before we continue we need to account for the fact that the class has been visited.
+        if (isDefault && isClassLike(symbol.getType())) visitedClassLikes.add(symbol.getName());
+        continue;
+      }
+      if (isStatic(symbol.getType()) && visitedClassLikes.contains(parentPath)) continue;
+
+      declareNamespace(isDefault ? parentPath : symbol.getName(), symbol, isDefault, compiler,
+          noTransitiveProvides, true);
+
+      if (!isDefault && isLikelyNamespace(symbol.getType(), compiler)) visitedNamespaces.add(symbol.getName());
+      if (isDefault && isClassLike(symbol.getType())) visitedClassLikes.add(symbol.getName());
+
       // we do not declare modules or goog.require support, because externs types should not be
       // visible from TS code.
     }
+  }
 
-    checkState(indent == 0, "indent must be zero after printing, but is %s", indent);
-    return out.toString();
+  private boolean isClassLike(JSType type) {
+    return isTrueConstructor(type) || type.isInterface();
```

Comment:
The name was confusing, renamed to 'allowsStaticFields' and indeed we do need to check for them separately, interfaces are not constructors.